File: geode-client-native/examples/clicache/ExecuteFunctions/ExecuteFunctions.cs
/*
 * The ExecuteFunction Example.
 */

// Use standard namespaces
using System;
using System.Threading;

// Use the GemFire namespace
using GemStone.GemFire.Cache;
using System.Collections.Generic;

namespace GemStone.GemFire.Cache.Examples
{
  // The Function Execution example.
  //customer result collector
  public class MyResultCollector : IResultCollector
  {
    private bool m_resultReady = false;
    private CacheableVector m_results = null;
    private int m_addResultCount = 0;
    private int m_getResultCount = 0;
    private int m_endResultCount = 0;

    public int GetAddResultCount()
    {
      return m_addResultCount ;
    }
    public int GetGetResultCount()
    {
      return m_getResultCount ;
    }
    public int GetEndResultCount()
    {
      return m_endResultCount ;
    }
    public MyResultCollector()
    {
      m_results = new CacheableVector();
    }
    public void AddResult(IGFSerializable result)
    {
      m_addResultCount++;
      CacheableArrayList rs = result as CacheableArrayList;
      for(int i = 0; i < rs.Count; i++)
      {
	m_results.Add(rs[i]);
      }
    }
    public IGFSerializable[] GetResult()
    {
      return GetResult(50);
    }
    public IGFSerializable[] GetResult(UInt32 timeout)
    {
      m_getResultCount++;
      if(m_resultReady == true)
      {
	return m_results.ToArray();
      }
      else 
      {
	for(int i=0; i < timeout; i++)
	{
	  Thread.Sleep(1000);
          if(m_resultReady == true)
          {
	    return m_results.ToArray();
          }
	}
	throw new FunctionExecutionException(
	           "Result is not ready, endResults callback is called before invoking getResult() method");
	
      }
    }
    public void EndResults()
    {
       m_endResultCount++;
       m_resultReady = true;
    }
    public void ClearResults()
    {
      m_results.Clear();
    }
  }

  class ExecuteFunctions
  {        
    private static string getFuncName = "MultiGetFunction";
    private static string getFuncIName = "MultiGetFunctionI";

    static void Main(string[] args)
    {
      try
      {
        // Create CacheFactory using the settings from the gfcpp.properties file by default.
        CacheFactory cacheFactory = CacheFactory.CreateCacheFactory();

        Cache cache = cacheFactory.SetSubscriptionEnabled(true)
                                  .AddServer("localhost", 40404)
                                  .AddServer("localhost", 50505)          
                                  .Create();

        Console.WriteLine("Created the GemFire Cache");

        Region region = cache.CreateRegionFactory(RegionShortcut.CACHING_PROXY)          
                             .Create("partition_region");

        Console.WriteLine("Created the partition_region.");

        for(int i=0; i < 34; i++)
        {
          region.Put("KEY--"+i, "VALUE--"+i);
        }
      
        IGFSerializable[] routingObj = new IGFSerializable[17];
        int j=0;
        for(int i=0; i < 34; i++)
        {
          if(i%2==0) continue;
          routingObj[j] = new CacheableString("KEY--"+i);
	  j++;
        }
        Console.WriteLine("routingObj count= {0}.", routingObj.Length);

        //test data dependant function execution
        //     test get function with result
        Boolean getResult = true;
        IGFSerializable args0 = new CacheableBoolean(true);
        Execution exc = FunctionService.OnRegion(region);
        IResultCollector rc =  exc.WithArgs(args0).WithFilter(routingObj).Execute(
	  getFuncName, getResult);
        IGFSerializable[] executeFunctionResult = rc.GetResult();
        Console.WriteLine("on region: result count= {0}.", executeFunctionResult.Length);

        List<IGFSerializable> resultList = new List<IGFSerializable>();

        for (int pos = 0; pos < executeFunctionResult.Length; pos++) {
          CacheableArrayList resultItem = executeFunctionResult[pos] as CacheableArrayList;
          foreach (IGFSerializable item in resultItem) {
            resultList.Add(item);
          }
        }
        Console.WriteLine("on region: result count= {0}.", resultList.Count);
        for (int i = 0; i < resultList.Count; i++) {
          Console.WriteLine("on region:get:result[{0}]={1}.", i, (resultList[i] as CacheableString).Value);
        }        
            
        getResult = true;
        //test date independant fucntion execution on one server
        //     test get function with result
        exc = FunctionService.OnServer(cache);
        CacheableVector args1 = new  CacheableVector();
        for(int i=0; i < routingObj.Length; i++)
        {
          Console.WriteLine("routingObj[{0}]={1}.", i, (routingObj[i] as CacheableString).Value);
          args1.Add(routingObj[i]);
        }
        rc =  exc.WithArgs(args1).Execute(
	  getFuncIName, getResult);
        executeFunctionResult = rc.GetResult();
        Console.WriteLine("on one server: result count= {0}.", executeFunctionResult.Length);

        List<IGFSerializable> resultList1 = new List<IGFSerializable>();
        for (int pos = 0; pos < executeFunctionResult.Length; pos++) {
          CacheableArrayList resultItem = executeFunctionResult[pos] as CacheableArrayList;
          foreach (IGFSerializable item in resultItem) {
            resultList1.Add(item);
          }
        }
        
        for (int i = 0; i < resultList1.Count; i++) {
          Console.WriteLine("on one server:get:result[{0}]={1}.", i, (resultList1[i] as CacheableString).Value);
        }        

        //test date independant fucntion execution on all servers
        //     test get function with result
        exc = FunctionService.OnServers(cache);
        rc =  exc.WithArgs(args1).Execute(getFuncIName, getResult);
        executeFunctionResult = rc.GetResult();
        Console.WriteLine("on all servers: result count= {0}.", executeFunctionResult.Length);

        List<IGFSerializable> resultList2 = new List<IGFSerializable>();
        for (int pos = 0; pos < executeFunctionResult.Length; pos++) {
          CacheableArrayList resultItem = executeFunctionResult[pos] as CacheableArrayList;
          foreach (IGFSerializable item in resultItem) {
            resultList2.Add(item);
          }
        }
        if (resultList2.Count != 34)
          Console.WriteLine("result count check failed on all servers");
        for (int i = 0; i < resultList2.Count; i++) {
          Console.WriteLine("on all servers:result[{0}]={1}.", i, (resultList2[i] as CacheableString).Value);
        }       

        //test withCollector
        MyResultCollector myRC = new MyResultCollector();
        rc =  exc.WithArgs(args1).WithCollector(myRC).Execute(getFuncIName, getResult);
        executeFunctionResult = rc.GetResult();
        Console.WriteLine("add result count= {0}.", myRC.GetAddResultCount());
        Console.WriteLine("get result count= {0}.", myRC.GetGetResultCount());
        Console.WriteLine("end result count= {0}.", myRC.GetEndResultCount());
        Console.WriteLine("on all servers with collector: result count= {0}.", executeFunctionResult.Length);

        // Close the GemFire Cache.
        cache.Close();

        Console.WriteLine("Closed the GemFire Cache");        
      }
      // An exception should not occur
      catch (GemFireException gfex)
      {
        Console.WriteLine("ExecuteFunctions GemFire Exception: {0}", gfex.Message);
      }
    }
  }
}

File: geode-client-native/quickstart/cpp/ExecuteFunctions.cpp
/*
 * The Execute Function QuickStart Example.
 *
 * This example takes the following steps:
 *
 * 1. Create a GemFire Cache.
 * 2. Create the example Region Programmatically.
 * 3. Populate some objects on the Region.
 * 4. Create Execute Objects
 * 5. Execute Functions
 * 6. Close the Cache.
 *
 */

// Include the GemFire library.
#include <gfcpp/GemfireCppCache.hpp>

// Use the "gemfire" namespace.
using namespace gemfire;

char* getFuncIName = (char*)"MultiGetFunctionI";
char* putFuncIName = (char*)"MultiPutFunctionI";
char* getFuncName = (char*)"MultiGetFunction";
char* putFuncName = (char*)"MultiPutFunction";


// The Execute Function QuickStart example.
int main(int argc, char ** argv)
{
  try
  {
    // Create CacheFactory using the settings from the gfcpp.properties file by default.
    CacheFactoryPtr cacheFactory = CacheFactory::createCacheFactory();

    LOGINFO("Created CacheFactory");
    
    // Create a GemFire Cache.
    CachePtr cachePtr = cacheFactory->setSubscriptionEnabled(true)->addServer("localhost", 50505)->addServer("localhost", 40404)->create();
    
    LOGINFO("Created the GemFire Cache");
    
    // Create the example Region Programmatically
    RegionPtr regPtr0 = cachePtr->createRegionFactory(CACHING_PROXY)->create("partition_region");
    
    LOGINFO("Created the Region");
    
    regPtr0->registerAllKeys();
    char buf[128];
    
    CacheableVectorPtr resultList = CacheableVector::create();
    for(int i=0; i < 34; i++)
    {
      sprintf(buf, "VALUE--%d", i);
      CacheablePtr value(CacheableString::create(buf));

      sprintf(buf, "KEY--%d", i);
      CacheableKeyPtr key = CacheableKey::create(buf);
      regPtr0->put(key, value);
    }

    bool getResult = true;
    CacheableVectorPtr routingObj = CacheableVector::create();
    for(int i=0; i < 34; i++)
    {
      if(i%2==0) continue;
      sprintf(buf, "KEY--%d", i);
      CacheableKeyPtr key = CacheableKey::create(buf);
      routingObj->push_back(key);
    }
    //test data independant function with result on one server
    LOGINFO("test data independant function with result on one server");
    CacheablePtr args = routingObj;
    ExecutionPtr exc = FunctionService::onServer((RegionServicePtr)cachePtr);
    CacheableVectorPtr executeFunctionResult = 
         exc->withArgs(args)->execute(getFuncIName, getResult)->getResult();
    if(executeFunctionResult==NULLPTR)
    {
      LOGINFO("get executeFunctionResult is NULL");
    } else 
    {
      for (int32_t item=0; item < executeFunctionResult->size(); item++)
      {
        CacheableArrayListPtr arrayList = dynCast<CacheableArrayListPtr>(executeFunctionResult->operator[](item));
        for (int32_t pos=0; pos < arrayList->size(); pos++)
        {
          resultList->push_back(arrayList->operator[](pos));
        }
      }
      sprintf(buf, "get: result count = %d", resultList->size());
      LOGINFO(buf);
      for(int32_t i=0; i < executeFunctionResult->size(); i++)
      {
         sprintf(buf, "get result[%d]=%s", i, dynCast<CacheableStringPtr>(resultList->operator[](i))->asChar());
         LOGINFO(buf);
      }
    }

    LOGINFO("test data independant function without result on one server");
    getResult = false;
    exc->withArgs(args)->execute(putFuncIName, getResult, 15, false);
    LOGINFO("test data independant function with result on all servers");
    getResult = true;
    exc = FunctionService::onServers((RegionServicePtr)cachePtr);
    executeFunctionResult = 
         exc->withArgs(args)->execute(getFuncIName, getResult)->getResult();
    if(executeFunctionResult==NULLPTR)
    {
      LOGINFO("get executeFunctionResult is NULL");
    } else 
    {
      resultList->clear();
      for (int32_t item=0; item < executeFunctionResult->size(); item++)
      {
        CacheableArrayListPtr arrayList = dynCast<CacheableArrayListPtr>(executeFunctionResult->operator[](item));
        for (int32_t pos=0; pos < arrayList->size(); pos++)
        {
          resultList->push_back(arrayList->operator[](pos));
        }
      }
      sprintf(buf, "get: result count = %d", resultList->size());
      LOGINFO(buf);
      for(int32_t i=0; i < executeFunctionResult->size(); i++)
      {
         sprintf(buf, "get result[%d]=%s", i, dynCast<CacheableStringPtr>(resultList->operator[](i))->asChar());
         LOGINFO(buf);
      }
    }

    getResult = false;
    LOGINFO("test data independant function without result on all servers");
    exc->withArgs(args)->execute(putFuncIName, getResult, 15, false);
    LOGINFO("test data dependant function with result");
    getResult = true;
    args = CacheableBoolean::create( 1 );
    exc = FunctionService::onRegion(regPtr0);
    executeFunctionResult = 
         exc->withFilter(routingObj)->withArgs(args)->execute(getFuncName, getResult)->getResult();
    if(executeFunctionResult==NULLPTR)
    {
      LOGINFO( "execute on region: executeFunctionResult is NULL");
    } else 
    {
      resultList->clear();
      LOGINFO( "Execute on Region: result count = %d", executeFunctionResult->size());
      for(int32_t i=0; i < executeFunctionResult->size(); i++)
      {
        CacheableArrayListPtr arrayList = dynCast<CacheableArrayListPtr>(executeFunctionResult->operator[](i));
        for (int32_t pos=0; pos < arrayList->size(); pos++)
        {
          resultList->push_back(arrayList->operator[](pos));
        }
      }
      sprintf(buf, "Execute on Region: result count = %d", resultList->size());
      LOGINFO(buf);
      
      for(int32_t i=0; i < resultList->size(); i++)
      {
         sprintf(buf, "Execute on Region: result[%d]=%s", i, dynCast<CacheableStringPtr>(resultList->operator[](i))->asChar());
         LOGINFO(buf);
      }
    }

    LOGINFO("test data dependant function without result");
    getResult = false;
    exc->withFilter(routingObj)->withArgs(args)->execute(putFuncName, getResult, 15, false);
    // Close the GemFire Cache.
    cachePtr->close();

    LOGINFO("Closed the GemFire Cache");
  }
  // An exception should not occur
  catch(const Exception & gemfireExcp)
  {
    LOGERROR("Function Execution GemFire Exception: %s", gemfireExcp.getMessage());
  }
}
